forem/forem
# Header navigation dropdown accessibility

## What type of PR is this? (check all applicable)


## Description

To make the header menu more ergonomic and accessible, this PR includes some updates to the markup and JavaScript logic. Including:

- Wrapping the logged-in header menu in an unordered list so the items are counted in a screen reader
- Adding `aria-expanded` to the navigation button so it indicates its state programmatically
- Opening the dropdown when you activate the button with a keyboard "click", as opposed to opening it on focus
- Sending focus to the first item when shown
- Wiring up the Escape key to close and sending focus back to the menu button

Note: the profile focus state could still use a little work, as it's not respecting border radius.

Because the navigation button also opens a logged-out overlay, I specifically did not add `aria-haspopup`, which is meant to indicate a sub menu. I considered adjusting the markup of that overlay to be an unordered list / menu-type-thing too, but it was going to be difficult with all of the wrapping DIVs and non-interactive items like heading and paragraph text. So I opted to leave that part alone.

A follow-up task will be to ensure this logic is repeated on all dropdown instances, starting with Storybook.

## Related Tickets & Documents

Related to https://github.com/forem/forem/issues/1154

## QA Instructions, Screenshots, Recordings

Navigate the header with the keyboard, using the Enter or Space key to open it and Escape to close (or tabbing past the last item, which I kept). Notice how you aren't forced through the open menu item anymore when you tab through the page.

Note: because the items are wrapped in list items now, I moved the logic to show the Forem Admin item to the LI instead of the A. This also eliminates an empty list item that was still being rendered when the Admin link was hidden for permissions reasons.

Tested in Voiceover and Safari, iOS Voiceover, Edge and NVDA, Chrome and NVDA, Firefox and NVDA.

## Added tests?

- [X] No, and this is why: We need Cypress for these tests! Alternatively I could rewrite this menu logic into an ES module and test it with Jest.

## Added to documentation?


## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![Multiasking with keyboards and coffee](https://media.giphy.com/media/Xbn8ZbO95YeOJQtJz7/giphy.gif)
